### ISSUE TITLE
Create fused matMul + relu shader for frozen model inference

### ISSUE THREAD

**Issue by annxingyuan** [tensorflow/tfjs#825]
When matMul and an activation e.g. relu are adjacent in a frozen model, encode them as a single new op: `matMulWithActivation`. This should happen only for inference so don't need to worry about gradients yet.

**Comment by astojilj** [comment]
> [Fuse matMul with activation and bias for performance. (#1475)](https://github.com/tensorflow/tfjs-core/commit/5a95a6407959c3b49675c9de8f013f1976df49ce) 

@annxingyuan, very nice. Thanks. This could be used to fold "conv2D -> BatchNorm -> Relu6" from MobileNetV2 to single operation: folding conv2D+BatchNorm to conv2D+BiasAdd (e.g. with [graph_transform](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#fold_batch_norms)) and then we execute this conv2D as MatMul...
Has the work on modifying tfjs-converter conversion to export fused nodes already started for this, or the plan is different?

**Comment by annxingyuan** [comment]
@astojilj 

We have no immediate plans for modifying tfjs-converter to export fused nodes as that would require graph rewriting. However we hope to release a patch for tfjs-layers that would take advantage of fused nodes - in this case no graph rewriting is needed because the layers API allows specifying `activation` / `bias` for different layer types (dense, convolutional, etc.). 

We definitely would like to invest more into fusing though - we're realizing more and more how underutilized our shaders are...

**Comment by astojilj** [comment]
@annxingyuan 
What's your opinion about approach where fusing would happen as part of nodes compiling in tfjs-converter - like in the [prototype here](https://github.com/tensorflow/tfjs-converter/compare/master...astojilj:fusedconvaddact#diff-91817df5c918eb884dea14ca23240852)?

There, I have started prototyping fused conv2d+biasAdd+relu6; this is [corresponding tfjs-core code](https://github.com/tensorflow/tfjs-core/compare/master...astojilj:fusedconvaddact?expand=1). 

DeepLabV3 MobileNetV2+ inference is showing very minor performance improvement although, from original 187 operations in the model, there are 15 fused poinwise conv2d+biasAdd+relu6. I plan to do benchmarking and try to fuse more conv2d+biasAdd pairs.

**Comment by annxingyuan** [comment]
@astojilj 

Very interesting! I'm impressed that you went down that rabbit hole with the converter :) 

However I think we should ideally avoid getting into the business of graph rewriting in the converter. I think the more sustainable approach is to work with the [Grappler](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/grappler) team to do this sort of work upstream. 

That being said, if there's a very compelling use case - i.e. a popular model being run through the converter shows *significant* improvements with fusing - we might need to reevaluate...

**Comment by astojilj** [comment]
@annxingyuan 

Thanks. 

I could [see that Grappler](https://github.com/tensorflow/tensorflow/commit/bd2610f638bd35dca30bb00a6c16da775b3bfcdf) is already working on it.
The approach I propose here is similar to what, my understanding is like that, Grappler is doing for tensorflow: in run time (during compilation) optimize graph for inference. In a way, I wouldn't use term rewrite.

Alternatively, this sort of approach could be offered through custom ops to developers.
I feel I'm hijacking this issue for something else. Let me branch to new issue and continue there...
